cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,266 @@
+# Resource Slack
+
+## Abstract
+
+"Resource slack" is some extra capacity in a k8s cluster kept for upcoming pods so that a newly created pod fit within resource slack does not need to wait until an extra node is started and gets ready.
+It enables pods to be scaled out faster while minimizing impact to resource utilization.
+This document sketches how the resource slack feature will be implemented in a future release of K8S.
+
+## Motivation
+
+Currently, ClusterAutoscaler tries to maintain necessary and sufficient amount of nodes.
+For scaling up, it adds nodes only after some unschedulable pods are observed.
+This behavior of scaling up implies that, in a worst case, an unschedulable pod must wait until an eligible node is created, started, and then gets ready before it is finally scheduled.
+A node takes a much longer time than that of a pod to start and become ready and therefore it results in pods to occasionally delayed for several minutes to be horizontally scaled out.
+A possible work-around can be done today is to lower the threshold of HPA to trigger a scaling up of a replicaset/deployment earlier but it reduces total resource utilization when there are many replicasets and deployments required the work-around.
+Having the resource slack feature allows us to scale out pods faster while minimizing impact to resource utilization.
+
+## Design
+
+To achieve Resource Slack, introduce placeholder pods that will live only inside of ClusterAutoscaler.
+They are created to virtually request certain amounts of CPU and memory so that they influence all of the scale-up/scale-down decision made in CA, but won't be present in the scheduler running inside an actual K8S cluster, so the new pods will be able to pick the place of the placeholders and push the placeholders elsewhere (and make them trigger the scale-up/down).
+
+### Requirements
+
+Version 1
+
+* [R1] Works on relatively large clusters today's CA used to work
+  * dozens of nodes, hundreds of pods
+  * In general we don't want to have O(pods * nodes) operations every single loop
+  * [R1-1] On large cluster there is a huge number of pods to be scheduled. And the pods are mostly the same. If a pod from one replica set X failed to schedule on node A then there is no point to in trying to fit the next pod from X on A. Thus there should be just a single loop over all nodes (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312)
+  * [R1-2] The scale down performance strongly depends that the pods don't change every iteration. So if you placed placeholder-X-123 on node A then in the next iteration it should also land on A or the scale down performance will be heavily impacted (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312)
```

Comment:
Will [R1-2] apply to scale-up as well? R1-1 is an optimization, but it's doesn't improve big-O complexity as far as I can tell. But if the placeholder pod keeps the information where it goes we only have to verify if it still fits where we placed it and only iterate over nodes if we can't.